PersistentVolume dynamic provisioning

markturansky opened the pull request:
Following PersistentVolumes, this proposal seeks to add the ability for PersistentVolumeControllers to maintain "replicas" of volumes, much like a ReplicationController maintains levels of pods.

PersistentVolumePlugins gain Create and Recycle methods to aid provisioning.

markturansky committed:
PersistentVolume dynamic provisioning

thockin commented on docs/design/persistent-volume-provisioning.md:
````diff
@@ -0,0 +1,129 @@
+
+The PersistentVolumeSource interface requires ```Create``` and ```Recycle``` functions to support the dynamic creation and reclamation of persistent volumes.  Each storage provider implements its own PersistentVolumeSource interface.
+
+Caveat:  Not all persistent volumes will support dynamic provisioning.
````
I assume this means things like NFS for which provisioning is not just an API call and is not consistent across installations.  But how can you make it possible?  Something like a PVSourceOverride that calls a custom script?  Maybe that's version 3.0

swagiaal replied:
I am not sure if this is what @markturansky intends but I think implementations of PersistentVolumeSource should be independent of the VolumePlugin implementations where it makes sense. For GCE PD for example they are one and the same, but for NFS there can be multiple PersistentVolumeSource implementors which result in NFS exports. There can also be a PersistentVolumeSource which can produce both NFS and iSCSI exports for example.
In the case of just plain NFS with no cloud provider a custom PVSource would make sense.

markturansky replied:
So long as a plugin can call exec and the admin creating the PVControllers can specify what to call, I suppose the caveat should be "some assembly required" for some types of volumes.

thockin commented on docs/design/persistent-volume-provisioning.md:
```diff
@@ -0,0 +1,129 @@
+// PersistentVolumeControllerSpec is the specification of a PV controller.
+// The controller creates volumes similar to the Template until the available count equals the replica count
+type PersistentVolumeControllerSpec struct {
+	// Replicas is the minimum number of available persistent volumes desired in the system
```
define "available" ?

markturansky replied:
unbound volumes are "available".

thockin commented on docs/design/persistent-volume-provisioning.md:
```diff
@@ -0,0 +1,129 @@
+
+Add/Sync events inspect for binding (pv.Spec.ClaimRef) and tallies available/bound.
+
+Uupdate events inspect for binding and phase.  Any volume unbound from its claim will be available for recycling.
```
Typo: Uupdate

thockin commented on docs/design/persistent-volume-provisioning.md:
````diff
@@ -0,0 +1,129 @@
+
+Two new persistent volume functions:
+
+The PersistentVolumeSource interface requires ```Create``` and ```Recycle``` functions to support the dynamic creation and reclamation of persistent volumes.  Each storage provider implements its own PersistentVolumeSource interface.
````
So replica count can go up but never down?  what do you do if you realize you asked for 10 replicas and you have 20?

markturansky replied:
I'm not necessarily a fan of the term "replica" here unless it is meaningful in a way I am unfamiliar with (entirely possible!).  I copied much of ReplicationController because these two concepts seem very similar.

"MinimumAvailable" is more aptly named in this context.

thockin replied:
But, and I'm making an assumption here, "number available" goes down when one is claimed.  This means that, over time you will have more and more volumes created and never deleted.  These things cost money.  I don't think that is tenable.

markturansky replied:
That is true.  If all claims got deleted, the number of available volumes might greatly exceed the "MinimumAvailable" count.  The pseudo-code for the controller contains a clause for recycling the excess volumes.  I suppose the difficulty lies in knowing what "recycle" means for some volumes.

For cloud volumes, deleting a volume for "recycling" seems like the best option.  That would decrease the available count.

For NFS, erasing a volume means it's still available and the loop that recycles the extras will loop forever and none get removed.

Let me think about this.  The plugin should probably advise the controller about what to do, since their "recycling" behavior is implementation specific.

markturansky replied:
I attempted to clarify the role of the PVCtrlMgr and how it would delete excess volumes.

thockin commented on docs/design/persistent-volume-provisioning.md:
````diff
@@ -0,0 +1,129 @@
+Uupdate events inspect for binding and phase.  Any volume unbound from its claim will be available for recycling.
+
+```
+if controller.replicas > 0:
````
This is not right - certainly you won't just blindly make new ones if the availableCount is sufficient beforehand

thockin commented on docs/design/persistent-volume-provisioning.md:
````diff
@@ -0,0 +1,129 @@
+        recycle(v)
+```
+
+NFS is a good example where controller.replicas will equal 0.  Dynamic creation will not take place, but the controller
````
So you really need to create 1 PersistentVolumeController per PVSource type in EVERY CLUSTER or else you will "leak" PVs.  And what selector do you apply for "manually created" PVs?

Can this aspect be streamlined?  Or do you not intend to auto-recycle manually created PVs?

markturansky replied:
I think all analogies to ReplicationController are valid.    Manually provisioned PVs would ideally fall under the management of a PersistentVolumeController as a "replica" (I don't necessarily want to steal this name, unless it is particularly meaningful in this context.  You tell me).

thockin replied:
This sort of ties in to my question about deleting replcas.  RC will delete pods if it is over-satisfied.

thockin commented on docs/design/persistent-volume-provisioning.md:
```diff
@@ -0,0 +1,129 @@
+
+#### Limitations
+
+The controller should not be mounting volumes, formatting disks, or other tasks associated with provisioning.  Ideally, a job
```
job queue means running as pods?

markturansky replied:
I thought about having a pod created in that namespace to perform some task (say, mount NFS and run "rm -rf" on the volume).

Is a pod running a task the same as a job that can report state?  Can I rely on the pods "Succeeded" phase to know it ran correctly in a namespace and it performed it's "rm -rf" task?

thockin replied:
You can either require the admin to run a pod and then you communicate with that pod or you can run pods yourself.  Running pods will not be able to format a disk, but they should be able to rm on it.

Our support for pods that complete is overall weak right now

erictune commented:
ok to test

thockin commented:
As simple as this "PR" is, I don't want to commit it because I simply have not put enough time into thinking about this model yet.

markturansky commented:
@thockin This proposal can wait until post 1.0.  Just the PRs for recycling are current.  We can revisit the rest of this later.

markturansky commented on docs/design/persistent-volume-provisioning.md:
```diff
@@ -0,0 +1,122 @@
+One new API kinds:
+
+A `PersistentVolumeController` (PVC) is a storage resource provisioned by an administrator.  PVCs do not have a namespace.  Like a ReplicationController maintains a minimum number of "replicas" of a pod in the system, so a PersistentVolumeController maintains a minimum number of "replicas" of a certain type/size of PersistentVolume available in the system.
+
```
@smarterclayton Having to make many ReplicationControllers with Replicas=1 and 1 claim is a bad user experience.

if a user were to lay claim to a `PersistentVolumeController` -- which is a creator and manager of its own pool of storage -- could we use this in conjunction with a ReplicationController to scale something like MongoDB? Different dev effort than this PR, but this PR's work could be steps towards solving that scaling issue.

smarterclayton replied:
On Jul 14, 2015, at 2:45 PM, Mark Turansky <notifications@github.com> wrote:

In docs/design/persistent-volume-provisioning.md
<https://github.com/GoogleCloudPlatform/kubernetes/pull/6773#discussion_r34603920>
:

> @@ -0,0 +1,122 @@
> +# Persistent Volume Provisioning
> +
> +This document proposes a model for dynamically provisioning [Persistent Volumes](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/design/persistent-storage.md)
> +
> +### tl;dr
> +
> +One new API kinds:
> +
> +A `PersistentVolumeController` (PVC) is a storage resource provisioned by an administrator.  PVCs do not have a namespace.  Like a ReplicationController maintains a minimum number of "replicas" of a pod in the system, so a PersistentVolumeController maintains a minimum number of "replicas" of a certain type/size of PersistentVolume available in the system.
> +

@smarterclayton <https://github.com/smarterclayton> Having to make many
ReplicationControllers with Replicas=1 and 1 claim is a bad user experience.

if a user were to lay claim to a PersistentVolumeController -- which is a
creator and manager of its own pool of storage -- could we use this in
conjunction with a ReplicationController to scale something like MongoDB?
Different dev effort than this PR, but this PR's work could be steps
towards solving that scaling issue.



Possible.  The design of how we assign unique volumes to pods under a
replication controller (and reuse them when pods dies) is really important,
so we should try and get the discussion moving forward on that and reach
closure.  Nominal services are tied to this.


—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/6773/files#r34603920>
.

markturansky commented:
I pushed edits to the doc reflecting the new implementation of the Recycler interface.

I can start the implementations of Deleter and Creator since they will follow Recycler.  No API changes required for those additions.

markturansky commented:
@smarterclayton this is a big focus for us at RH doing storage.

markturansky commented:
@thockin Any reason I cannot start the implementation for this design doc?  Even if we tweak it, would it be close enough so that getting started wouldn't involve a lot of rework?

bgrant0607 commented on docs/design/persistent-volume-provisioning.md:
```diff
@@ -0,0 +1,140 @@
+
+One new API kind:
+
+A `PersistentVolumeController` (PVC) is a storage resource provisioned by an administrator.  PVCs do not have a namespace.  Just as a `ReplicationController` maintains a number of replicas of a pod, a `PersistentVolumeController` maintains a minimum number of replicas of a `PersistentVolume`.  A PVC creates new volumes from a template up to a maximum replica count in increments of the minimum rep ...
```
PVC is the abbreviation we use for PersistentVolumeClaim.

bgrant0607 replied:
This replicates PVs or PVClaims?

If the former, why isn't this a PV auto-scaler?

bgrant0607 replied:
cc auto-scaling folks
@davidopp @jszczepkowski

davidopp replied:
I read this as replicating PVs, not PVClaims. The Spec describes a PV.

I can kinda see the analogy to auto-scaler but I think it's closer to a simple ReplicationController, with the twist that there is an upper bound on the total number of PVs in any phase in addition to a goal number of PVs in Available phase (while RC only cares about Pods in Running phase). I didn't get the sense that it is dynamically adjusting the number of PVs to try to keep some application-level metric within a particular range, so I don't think it is really an auto-scaler.

markturansky replied:
@davidopp has it exactly right.  PVCtrl is the equivalent to a ReplicationController with min/max numbers and David's observation about the Available phase.

davidopp commented on docs/design/persistent-volume-provisioning.md:
```diff
@@ -0,0 +1,140 @@
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+# Persistent Volume Provisioning
+
+This document proposes a model for dynamically provisioning [Persistent Volumes](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/persistent-volumes.md)
```
I think this URL should be
https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/persistent-volumes.md

davidopp commented on docs/design/persistent-volume-provisioning.md:
````diff
@@ -0,0 +1,140 @@
+
+```
+
+Dynamic pools of storage (e.g, GCE) will maintain `MinimumReplicas` of a `PersistentVolume` at a given time where each PVs status phase is ` Available`, up to `MaximumReplicas`.  Storage levels will increase by one until the minimum or maximum levels are reached.  Recycling a PV means the current number of available volumes might exceed the minimum replica count.  Plugins implementing the Deleter ...
````
In the first sentence, make it clearer that MaximumReplicas is the maximum number of replicas in any phase, not just Available.

davidopp commented on docs/design/persistent-volume-provisioning.md:
````diff
@@ -0,0 +1,140 @@
+
+Dynamic pools of storage (e.g, GCE) will maintain `MinimumReplicas` of a `PersistentVolume` at a given time where each PVs status phase is ` Available`, up to `MaximumReplicas`.  Storage levels will increase by one until the minimum or maximum levels are reached.  Recycling a PV means the current number of available volumes might exceed the minimum replica count.  Plugins implementing the Deleter ...
+
+Static pools of storage (e.g, NFS) will have ```MinimumReplicas``` == ```MaximumReplicas```.  Creation or deletion of the resource will not occur but the controller allows recycling of the volume through the plugin.
````
Don't we already have a mechanism for recycling a volume without adding a new controller?
https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/persistent-volumes.md#recycling-policy

(In other words, can we avoid having a PV Controller for static pools?)

markturansky replied:
Yes, the current recycler control loop will work just fine for the Delete reclaim policy.  I originally had it in there pre-1.0 but removed it until the rest of this design was ready to be implemented.

We need to add 1 controller for the PersistentVolumeControllerManager which will handle the PVCtrls and creation of PVs.

I'll update this doc to reflect the current design, too.  This proposal was made pre-1.0 and only recycling was able to fit at the very end.

davidopp commented on docs/design/persistent-volume-provisioning.md:
````diff
@@ -0,0 +1,140 @@
+** PVs matching the selector but made manually count as replicas if they match the label selector
+** PVs made manually or from earlier templates may have difference size capacities.  Delete largest first.
+* create a new volume ``` if unboundReplicas < minimumReplicas && totalReplicas <= maximumReplicas``` 
+* delete a volume ``` while totalReplicas > maximumReplicas, delete largest volume ```
````
would you ever delete a bound replica? it seems you could get into that situation if someone manually creates a volume, outside of the controller, with a matching label.

davidopp commented on docs/design/persistent-volume-provisioning.md:
```diff
@@ -0,0 +1,140 @@
+| gluster  | No, externally provisioned | pre-formatted during provisioning | mount volume, run "rm -rf" on volume | ? | ? | ? |
+| ceph  | No, externally provisioned | pre-formatted during provisioning | mount volume, run "rm -rf" on volume | ? | ? | ? |
+
+#### Recycler implementation 
```
See earlier comment about being confused, I thought we already support volume recycling.

smarterclayton commented on docs/design/persistent-volume-provisioning.md:
```diff
@@ -0,0 +1,140 @@
+### Goals
+
+* Allow administrators to describe minimum and maximum storage levels comprised of many kinds of PersistentVolumes
+* Allow the dynamic creation and reclamation of persistent volumes (to the fullest extent for each type of storage provider)
```
How does the system allow for this?  My question about this design proposal is that it seems to imply that the easiest way to provision volume types is to write Go code that calls GCE apis.  That works well specifically for GCE/AWS/Cinder (in the normal case) but does not work very well for any organization that has to script their storage creation (by using ansible or some other tool to provision new machines or carve up NFS mounts).  It also doesn't work very well for users with custom needs on GCE, or the ability to dynamically react to a claim and create a volume on demand.  This seems like a hard problem to generically solve (provisioning storage of many different types) such that I'd ask why it has to be implemented in a formal pattern, vs by an integration that someone writes to watch for new claims or a lack of volumes and go create some more.

pmorie replied:
I'd like to be able to run a pod on the server hosting the volumes (assuming it is itself a node in the cluster) in order to create a volume.

smarterclayton replied:
Why does it have to be on the server hosting the volumes?

On Thu, Aug 20, 2015 at 3:30 PM, Paul Morie <notifications@github.com>
wrote:

> In docs/design/persistent-volume-provisioning.md
> <https://github.com/kubernetes/kubernetes/pull/6773#discussion_r37571168>:
>
> > +
> > +`PersistentVolumeControllerManager` is a singleton control loop running in master that manages all PVControllers in the system.  The PVCM reconciles the current supply of available PersistentVolumes in the system with the desired levels according to the PVControllers.   This process is similar to the ```ReplicationManager``` that manages ReplicationControllers.
> > +
> > +Three new volume plugin interfaces:
> > +
> > +* Recycler -- knows how to scrub a volume clean so it can become available again as a resource
> > +* Creator -- create new instances of a PV from a template.
> > +* Deleter -- deletes instances of a PV and allows the plugin to determine how to remove it from the underlying infrastructure
> > +
> > +Volume plugins can implement any applicable interfaces.  Each plugin will document its own support for dynamic provisioning.
> > +
> > +
> > +### Goals
> > +
> > +* Allow administrators to describe minimum and maximum storage levels comprised of many kinds of PersistentVolumes
> > +* Allow the dynamic creation and reclamation of persistent volumes (to the fullest extent for each type of storage provider)
>
> I'd like to be able to run a pod on the server hosting the volumes
> (assuming it is itself a node in the cluster) in order to create a volume.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/6773/files#r37571168>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

markturansky replied:
@smarterclayton yes, this works much more easily for volumes w/ APIs.  

I was thinking about making the strategy for creation be plugin-based.  Perhaps the strategies are compiled in the /plugins package and you choose one by name via config (with a sensible default, of course).

If we made creation pluggable, @pmorie can run his pod to create a volume or Go code can call a provider's API.

thockin replied:
@smarterclayton Are you arguing that this doesn't need API objects, but could just be a super-privileged pod that knows policy and watches across all namespaces for unfulfilled claims and does whatever it needs to do to make new PVs?

This whole design-doc sort of parallels the work @bprashanth is doing on load-balancers, so I'd like to keep them similar in form.

@saad-ali also

smarterclayton replied:
I think the key distinction is that the in tree controller would only
support a few obviously core types (like a cloud provider).  A custom
controller would not be in tree, or compiled with the manager, and we would
disable it.

On Thu, Aug 20, 2015 at 8:46 PM, Mark Turansky <notifications@github.com>
wrote:

> In docs/design/persistent-volume-provisioning.md
> <https://github.com/kubernetes/kubernetes/pull/6773#discussion_r37597422>:
>
> > +
> > +`PersistentVolumeControllerManager` is a singleton control loop running in master that manages all PVControllers in the system.  The PVCM reconciles the current supply of available PersistentVolumes in the system with the desired levels according to the PVControllers.   This process is similar to the ```ReplicationManager``` that manages ReplicationControllers.
> > +
> > +Three new volume plugin interfaces:
> > +
> > +* Recycler -- knows how to scrub a volume clean so it can become available again as a resource
> > +* Creator -- create new instances of a PV from a template.
> > +* Deleter -- deletes instances of a PV and allows the plugin to determine how to remove it from the underlying infrastructure
> > +
> > +Volume plugins can implement any applicable interfaces.  Each plugin will document its own support for dynamic provisioning.
> > +
> > +
> > +### Goals
> > +
> > +* Allow administrators to describe minimum and maximum storage levels comprised of many kinds of PersistentVolumes
> > +* Allow the dynamic creation and reclamation of persistent volumes (to the fullest extent for each type of storage provider)
>
> @smarterclayton <https://github.com/smarterclayton> yes, this works much
> more easily for volumes w/ APIs.
>
> I was thinking about making the strategy for creation be plugin-based.
> Perhaps the strategies are compiled in the /plugins package and you choose
> one by name via config (with a sensible default, of course).
>
> If we made creation pluggable, @pmorie <https://github.com/pmorie> can
> run his pod to create a volume or Go code can call a provider's API.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/6773/files#r37597422>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

bgrant0607 replied:
@thockin I'm arguing exactly what you said here: https://github.com/kubernetes/kubernetes/pull/6773/files#r37607582

PVs are analogous to Nodes. We don't have NodeSets. We don't have CloudLoadBalancerSets. We shouldn't need PersistentVolumeSets. Create PVs on demand in response to PVCs. One could also potentially horizontally auto-scale to keep a small amount of burst capacity.

thockin replied:
Devil's advocate - will we ever have some configuration for a k8s cluster that defines the minimum number of nodes, the maximum number of nodes, the distribution of node sizes for auto-creation, the threshold for auto-creation/deletion, and hysteresis?  We actually DO have this config but it isn't store in kubernetes - it's in the cloud auto-scaler.

A PVSet (bad name) as proposed here is the analog to that.  We could implement this as a pod with no state in the API, but the state has to go SOMEWHERE.  We could argue that this should be a config object, once that exists and in the mean time just expect cmdline flags on a pod (or secrets :) 

And then there is network ingress.  I think ingress follows the same pattern.  @bprashanth

bgrant0607 replied:
If we do ever add such types, we should involve the auto-scaling team.

markturansky commented:
@davidopp thanks for the feedback.  I addressed your comments in the doc and pushed a new revision.

This proposal was pre-1.0 and there was only time for the Recycler, hence some of the confusing references in the doc.  I added a little bit *more* on the Recycler so that this design doc is current and reflects the totality of the feature.  The delta left for implementation are Deleters, Creators, and the PVControllerManager loop.  I've got a PR with the API/Client for PVControllers.

thockin commented on docs/design/persistent-volume-provisioning.md:
```diff
@@ -0,0 +1,169 @@
+
+One new API kind:
+
+A `PersistentVolumeController` (PVCtrl) is a storage resource provisioned by an administrator.  PVCtrls do not have a namespace.  Just as a `ReplicationController` maintains a number of replicas of a pod, a `PersistentVolumeController` maintains a minimum number of replicas of a `PersistentVolume`.  A PVCtrl creates new volumes from a template up to a maximum replica count in increments of the mi ...
```
There's a desire to move from "active" names to "passive" names.  PersistentVolumePolicy?  PVPool?  PVSet?

I think @bgrant0607  wants to rename replicationController -> replicaSet eventually..

thockin replied:
I'm not sure I buy the last sentence.

markturansky replied:
I can change to `PersistentVolumeSet` if that's where RC is headed.  This thing makes a set of PVs.

I think I meant "PVs" in that last sentence, but that sentence also provides 0 value to the design doc because it may not be always true. I'll remove it.

markturansky commented:
@kubernetes/rh-storage @smarterclayton  @thockin @saad-ali  Is there value in PersistentVolumeSets aside from dynamic provisioning?  

Recycling exists for all storage that has no API.  I added support for Delete (#13649) and Create (#13650).

Do storage providers with APIs really need pre-provisioned resources or do they benefit more from the on-demand model?

I can add Conditions to PVClaims and the existing binder can optionally assign a "Provisionable" condition to unmatched PVCs.  This can all be configured via the new VolumeConfig.

A new controller watches for provisionable PVCs and uses the Creater interface.  

That's just two TODOs (Conditions and a new control loop) instead of a new top-level object and all the support it requires.

smarterclayton commented:
On Sep 7, 2015, at 4:32 PM, Mark Turansky <notifications@github.com> wrote:

@kubernetes/rh-storage <https://github.com/orgs/kubernetes/teams/rh-storage>
@smarterclayton <https://github.com/smarterclayton> @thockin
<https://github.com/thockin> @saad-ali <https://github.com/saad-ali> Is
there value in PersistentVolumeSets aside from dynamic provisioning?

Recycling exists for all storage that has no API. I added support for
Delete (#13649 <https://github.com/kubernetes/kubernetes/pull/13649>) and
Create (#13650 <https://github.com/kubernetes/kubernetes/pull/13650>).

Do storage providers with APIs really need pre-provisioned resources or do
they benefit more from the on-demand model?


My opinion is the latter more.


I can add Conditions to PVClaims and the existing binder can optionally
assign a "Provisionable" condition to unmatched PVCs. This can all be
configured via the new VolumeConfig.



Why do you need this step?  An unmatched pvc is implicitly provisionable


A new controller watches for provisionable PVCs and uses the Creater
interface.



Why wouldn't it just use the existing PV API to create a PV?


That's just two TODOs (Conditions and a new control loop) instead of a new
top-level object and all the support it requires.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/6773#issuecomment-138373236>.

smarterclayton commented:
I think that we don't want volume code for creating new volumes to be
invoked by a single controller.  Instead, each "provisioner" is its own
controller that watches pvcs and provisions them.  Creating a controller
loop is not supposed to be hard, and this is the classic "I don't like your
provision code, I'm going to write my own" case for an admin.

The controller pattern is our plugin pattern for these things.


On Sep 7, 2015, at 4:32 PM, Mark Turansky <notifications@github.com> wrote:

@kubernetes/rh-storage <https://github.com/orgs/kubernetes/teams/rh-storage>
@smarterclayton <https://github.com/smarterclayton> @thockin
<https://github.com/thockin> @saad-ali <https://github.com/saad-ali> Is
there value in PersistentVolumeSets aside from dynamic provisioning?

Recycling exists for all storage that has no API. I added support for
Delete (#13649 <https://github.com/kubernetes/kubernetes/pull/13649>) and
Create (#13650 <https://github.com/kubernetes/kubernetes/pull/13650>).

Do storage providers with APIs really need pre-provisioned resources or do
they benefit more from the on-demand model?

I can add Conditions to PVClaims and the existing binder can optionally
assign a "Provisionable" condition to unmatched PVCs. This can all be
configured via the new VolumeConfig.

A new controller watches for provisionable PVCs and uses the Creater
interface.

That's just two TODOs (Conditions and a new control loop) instead of a new
top-level object and all the support it requires.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/6773#issuecomment-138373236>.

markturansky commented:
You're right, I don't need a condition.  The pvc is "Pending" and unbound.  That's all a watcher needs to know to trigger a new resource.

I prefer nixing PersistentVolumeSet.  We can achieve the same functionality without an additional top level thing. The new volume config stuff allows better configuration of plugins.

The only candidates for implementation were AWS/GCE/OpenStack, which are the very ones we would want real dynamic provisioning for.  The existing recycler is good enough for statically provisioned pools of stoage.

smarterclayton commented:
Let's talk tomorrow and I'll go over the cli controller work so you can get
an idea of what we could offer if someone wants to write their own dynamic
provisioner or recycler.

On Sep 7, 2015, at 7:25 PM, Mark Turansky <notifications@github.com> wrote:

You're right, I don't need a condition. The pvc is "Pending" and unbound.
That's all a watcher needs to know to trigger a new resource.

I prefer nixing PersistentVolumeSet. We can achieve the same functionality
without an additional top level thing. The new volume config stuff allows
better configuration of plugins.

The only candidates for implementation were AWS/GCE/OpenStack, which are
the very ones we would want real dynamic provisioning for. The existing
recycler is good enough for statically provisioned pools of stoage.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/6773#issuecomment-138392440>.

markturansky commented:
Sounds good.  I will follow up in the morning.

bgrant0607 commented:
I haven't had time to look at this, sadly, but is there a reason why this should be distinct from nominal services #260? What if a pod has more than one persistent volume? I doubt the user would want arbitrary mix-and-match of volumes every time a pod were replaced.

saad-ali commented:
I'm a little late here I realize, but I'm trying to wrap my head around what the UX will look like for an admin to create a new dynamically provisioned volume set.

The PersistentVolumeTemplateSpec only contains PersistentVolumeSpec, will that be enough information to create a new volume for all volume types? For example how would one define the region for a GCE PD?

I imagine this is the type of detail  would be handled by the volume plugin, but how would the admin configure the plugin?

bgrant0607 commented on docs/design/persistent-volume-provisioning.md:
```diff
@@ -0,0 +1,169 @@
+
+One new API kind:
+
+A `PersistentVolumeSet` (PVS) is a storage resource provisioned by an administrator.  PVSets do not have a namespace.  Just as a `ReplicationController` maintains a number of replicas of a pod, a `PersistentVolumeSet` maintains a minimum number of replicas of a `PersistentVolume`.  A PVSet creates new volumes from a template up to a maximum replica count.
```
Observation, before having read the whole proposal:

PersistentVolumes are like Nodes -- they represent infrastructure, either provisioned by an admin on bare metal or potentially horizontally auto-scaled or provisioned on demand on a public/private cloud. It doesn't seem like there will be a uniform implementation across cloud providers. We don't have NodeSets partly for this reason.

PersistentVolumeClaims are analogous to Pods. We need replication of PVCs, but not independently of the Pods that consume them.  That's discussed in #260.

davidopp replied:
Can you explain the objection to PersistentVolumeSet? If they're all provisioned from the same template, it seems like the name correctly parallels ReplicaSet.

davidopp replied:
Also, I didn't understand the comment about PVCs. IIUC they're proposing replication of PersistentVolumes here, not replication of PVCs.

thockin commented:
given how inaccurate this doc is - should we close the PR?

markturansky commented:
Yes, we can close this PR.  The interfaces described are still relevant,
but can be described in the existing PV docs.

On Thu, Oct 1, 2015 at 4:28 PM, Tim Hockin <notifications@github.com> wrote:

> given how inaccurate this doc is - should we close the PR?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/6773#issuecomment-144838244>
> .
>
